File: py2_energy_test_final_feasible/minimum-stability-factor-of-array/minimum-stability-factor-of-array_benchmark.py
# coding: utf-8
import time

# ========== PASSING PYTHON 2 SOLUTION (100% UNMODIFIED) ==========
# Time:  O(nlogn * logr)
# Space: O(nlogn)

# number theory, binary search, rmq, sparse table, greedy
class Solution(object):
    def minStable(self, nums, maxC):
        """
        :type nums: List[int]
        :type maxC: int
        :rtype: int
        """
        if not nums:
            return 0
        def gcd(a, b):
            while b:
                a, b = b, a%b
            return a

        def binary_search_right(left, right, check):
            while left <= right:
                mid = left + (right-left)//2
                if not check(mid):
                    right = mid-1
                else:
                    left = mid+1
            return right

        # RMQ - Sparse Table
        # Template: https://github.com/kamyu104/GoogleCodeJam-Farewell-Rounds/blob/main/Round%20D/genetic_sequences2.py3
        # Time:  ctor:  O(NlogN) * O(fn)
        #        query: O(fn)
        # Space: O(NlogN)
        class SparseTable(object):
            def __init__(self, arr, fn):
                self.fn = fn
                self.bit_length = [0]
                n = len(arr)
                k = n.bit_length()-1  # log2_floor(n)
                for i in xrange(k+1):
                    self.bit_length.extend(i+1 for _ in xrange(min(1<<i, (n+1)-len(self.bit_length))))
                self.st = [[0]*n for _ in xrange(k+1)]
                self.st[0] = arr[:]
                for i in xrange(1, k+1):  # Time: O(NlogN) * O(fn)
                    for j in xrange((n-(1<<i))+1):
                        self.st[i][j] = fn(self.st[i-1][j], self.st[i-1][j+(1<<(i-1))])
        
            def query(self, L, R):  # Time: O(fn)
                i = self.bit_length[R-L+1]-1  # log2_floor(R-L+1)
                return self.fn(self.st[i][L], self.st[i][R-(1<<i)+1])

        def check(l):
            cnt = 0
            i = 0
            while i+l-1 < len(nums):
                if rmq.query(i, i+l-1) >= 2:
                    cnt += 1
                    i += l
                else:
                    i += 1
            return cnt > maxC

        rmq = SparseTable(nums, gcd)
        return binary_search_right(1, len(nums), check)

# ==================================================================

# ========== TEST HARNESS ==========
def main():
    iterations = 1000

    # Test cases extracted from C++ (convert C++ syntax to Python 2)
    arr9 = []
    for i in xrange(1, 51):
        if i % 5 == 0:
            arr9.append(0)
        elif i % 2 == 0:
            arr9.append(i * 2)
        elif i % 3 == 0:
            arr9.append(i * 3)
        else:
            arr9.append(i * 7 + 1)

    test_cases = [
        ([], 0),  # 0) Empty array
        ([2], 0),  # 1) Single element
        ([2, 4, 6, 8, 10, 12, 14, 16], 2),  # 2) All evens (high gcd)
        ([3, 5, 7, 11, 13, 17, 19], 0),  # 3) All odd primes (mostly gcd 1)
        ([0, 0, 2, 4, 0, 6, 0, 8, 10, 0], 3),  # 4) Mix with zeros
        ([-2, -4, 6, -8, 10, -12], 1),  # 5) Includes negatives
        ([14, 25, 9, 35, 22, 49, 18, 55, 26], 1),  # 6) Mixed values with likely gcd 1
        ([1, 1, 1, 1, 1, 1, 1], 5),  # 7) All ones (gcd always 1)
        ([12, 18, 24, 30, 36, 42, 48, 54], 5),  # 8) Composite numbers with many common factors
        (arr9, 7),  # 9) Larger mixed array programmatically generated
    ]

    # Warm-up
    sol = Solution()
    for test in test_cases:
        if isinstance(test, tuple):
            sol.minStable(*test)
        else:
            sol.minStable(test)

    # Benchmark
    results = []
    start_time = time.time()

    for iter_idx in xrange(iterations):
        sol = Solution()
        for test in test_cases:
            if isinstance(test, tuple):
                nums, maxC = test
                result = sol.minStable(nums, maxC + (iter_idx & 1))
            else:
                result = sol.minStable(test)
            results.append(result)

    elapsed_time = time.time() - start_time
    checksum = sum(hash(str(r)) for r in results) % (10**9 + 7)
    print("Checksum: {}, Total time: {:.6f}s".format(checksum, elapsed_time))

if __name__ == "__main__":
    main()

File: transpiled_after/minimum-stability-factor-of-array/cpp/minimum-stability-factor-of-array_benchmark.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <unordered_map>
#include <unordered_set>
#include <queue>
#include <stack>
#include <string>
#include <climits>
#include <cmath>
#include <numeric>
#include <chrono>

using namespace std;
using namespace std::chrono;

class Solution {
public:
    struct SparseTableGCD {
        vector<vector<int>> st;
        vector<int> log2of;
        SparseTableGCD() {}
        explicit SparseTableGCD(const vector<int>& arr) {
            int n = arr.size();
            log2of.assign(n + 1, 0);
            for (int i = 2; i <= n; ++i) {
                log2of[i] = log2of[i / 2] + 1;
            }
            int k = (n > 0) ? log2of[n] : 0;
            st.assign(k + 1, vector<int>(n, 0));
            if (n > 0) {
                st[0] = arr;
                for (int i = 1; i <= k; ++i) {
                    int len = 1 << i;
                    int half = len >> 1;
                    for (int j = 0; j + len <= n; ++j) {
                        st[i][j] = std::gcd(st[i - 1][j], st[i - 1][j + half]);
                    }
                }
            }
        }
        int query(int L, int R) const {
            if (L > R) return 0;
            int i = log2of[R - L + 1];
            return std::gcd(st[i][L], st[i][R - (1 << i) + 1]);
        }
    };

    int minStable(vector<int>& nums, int maxC) {
        int n = nums.size();
        SparseTableGCD rmq(nums);
        auto check = [&](int l) -> bool {
            long long cnt = 0;
            int i = 0;
            while (i + l - 1 < n) {
                if (rmq.query(i, i + l - 1) >= 2) {
                    ++cnt;
                    i += l;
                } else {
                    ++i;
                }
            }
            return cnt > maxC;
        };

        int l = 1, r = n;
        while (l <= r) {
            int mid = l + (r - l) / 2;
            if (!check(mid)) {
                r = mid - 1;
            } else {
                l = mid + 1;
            }
        }
        return r;
    }
};

struct TestCase {
    vector<int> nums;
    int maxC;
};

int main() {
    // Define 10 diverse test inputs
    vector<TestCase> tests;

    // 0) Empty array
    tests.push_back({{}, 0});

    // 1) Single element
    tests.push_back({{2}, 0});

    // 2) All evens (high gcd)
    tests.push_back({{2, 4, 6, 8, 10, 12, 14, 16}, 2});

    // 3) All odd primes (mostly gcd 1)
    tests.push_back({{3, 5, 7, 11, 13, 17, 19}, 0});

    // 4) Mix with zeros
    tests.push_back({{0, 0, 2, 4, 0, 6, 0, 8, 10, 0}, 3});

    // 5) Includes negatives
    tests.push_back({{-2, -4, 6, -8, 10, -12}, 1});

    // 6) Mixed values with likely gcd 1
    tests.push_back({{14, 25, 9, 35, 22, 49, 18, 55, 26}, 1});

    // 7) All ones (gcd always 1)
    tests.push_back({{1, 1, 1, 1, 1, 1, 1}, 5});

    // 8) Composite numbers with many common factors
    tests.push_back({{12, 18, 24, 30, 36, 42, 48, 54}, 5});

    // 9) Larger mixed array programmatically generated
    vector<int> arr9;
    for (int i = 1; i <= 50; ++i) {
        if (i % 5 == 0) arr9.push_back(0);               // introduce zeros
        else if (i % 2 == 0) arr9.push_back(i * 2);      // even numbers
        else if (i % 3 == 0) arr9.push_back(i * 3);      // multiples of 3
        else arr9.push_back(i * 7 + 1);                  // numbers often coprime
    }
    tests.push_back({arr9, 7});

    const int iterations = 1000;

    Solution solver;
    long long checksum = 0;

    auto start = high_resolution_clock::now();

    for (int iter = 0; iter < iterations; ++iter) {
        for (size_t k = 0; k < tests.size(); ++k) {
            // Slightly vary maxC per iteration to avoid being trivially optimized
            int maxC = tests[k].maxC + (iter & 1);
            checksum += solver.minStable(tests[k].nums, maxC);
        }
    }

    auto end = high_resolution_clock::now();
    auto dur_ns = duration_cast<nanoseconds>(end - start).count();
    double dur_ms = dur_ns / 1e6;

    // Print result to prevent optimization and report timing
    cout << "Checksum: " << checksum << '\n';
    cout << "Time: " << dur_ns << " ns (" << dur_ms << " ms)" << '\n';

    return 0;
}